crates/core/src/host_component.rs

```diff
@@ -0,0 +1,217 @@
+use std::{any::Any, marker::PhantomData, sync::Arc};
+
+use anyhow::Result;
+
+use super::{Data, Linker};
+
+/// A trait for Spin "host components".
+///
+/// A Spin host component is an interface provided to Spin components that is
+/// implemented by the host. This trait is designed to be compatible with
+/// [`wit-bindgen`](https://github.com/bytecodealliance/wit-bindgen)'s
+/// generated bindings.
+///
+/// # Example
+///
+/// ```ignore
+/// wit_bindgen_wasmtime::export!({paths: ["my-interface.wit"], async: *});
+///
+/// #[derive(Default)]
+/// struct MyHostComponent {
+///     // ...
+/// }
+///
+/// #[async_trait]
+/// impl my_interface::MyInterface for MyHostComponent {
+///     // ...
+/// }
+///
+/// impl HostComponent for MyHostComponent {
+///     type Data = Self;
+///
+///     fn add_to_linker<T: Send>(
+///         linker: &mut Linker<T>,
+///         get: impl Fn(&mut spin_core::Data<T>) -> &mut Self::Data + Send + Sync + Copy + 'static,
+///     ) -> anyhow::Result<()> {
+///         my_interface::add_to_linker(linker, get)
+///     }
+///
+///     fn build_data(&self) -> Self::Data {
+///         Default::default()
+///     }
+/// }
+/// ```
+pub trait HostComponent: Send + Sync + 'static {
+    /// Host component runtime data.
+    type Data: Send + Sized + 'static;
+
+    /// Add this component to the given Linker, using the given runtime state-getting handle.
+    ///
+    /// This function signature mirrors those generated by `wit-bindgen`.
+    fn add_to_linker<T: Send>(
+        linker: &mut Linker<T>,
+        get: impl Fn(&mut Data<T>) -> &mut Self::Data + Send + Sync + Copy + 'static,
+    ) -> Result<()>;
+
+    /// Builds new host component runtime data for [`HostComponentsData`].
+    fn build_data(&self) -> Self::Data;
+}
+
+impl<HC: HostComponent> HostComponent for Arc<HC> {
+    type Data = HC::Data;
+
+    fn add_to_linker<T: Send>(
+        linker: &mut Linker<T>,
+        get: impl Fn(&mut Data<T>) -> &mut Self::Data + Send + Sync + Copy + 'static,
+    ) -> Result<()> {
+        HC::add_to_linker(linker, get)
+    }
+
+    fn build_data(&self) -> Self::Data {
+        (**self).build_data()
+    }
+}
+
+/// An opaque handle returned by [`crate::EngineBuilder::add_host_component`]
+/// which can be passed to [`HostComponentsData`] to access or set associated
+/// [`HostComponent::Data`].
+pub struct HostComponentDataHandle<HC: HostComponent> {
+    idx: usize,
+    _phantom: PhantomData<fn() -> HC::Data>,
```

Comment:
I have not used `PhantomData` before so i am trying to get a sense of what it is being used for here. Is this helping map the data types for different languages for a host component?

Reply:
Rust requires that generic params be used in type definitions. In this case the struct represents a handle to some `HostComponent::Data` but doesn't actually hold any; `PhantomData` is a zero-sized placeholder to appease the type checker.